pennylane/compiler/python_compiler/transforms/quantum/diagonalize_measurements.py

```diff
@@ -102,6 +102,18 @@ def match_and_rewrite(self, observable: NamedObsOp, rewriter: pattern_rewriter.P
 
                 qubit = gate.out_qubits[0]
 
+            # we need to replace the initial qubit use everwhere EXCEPT the use that is now the
+            # input to the first diagonalizing gate. Its not enough to only change the NamedObsOp,
+            # because the qubit might be inserted/deallocated later
+            uses_to_change = [
+                use for use in observable.qubit.uses if not isinstance(use.operation, CustomOp)
```

Comment:
Hypothetically, the operation could be any gate right? Is this overkill? 😅 
```suggestion
                use for use in observable.qubit.uses if not isinstance(use.operation, (CustomOp, GlobalPhaseOp, MultiRZOp, QubitUnitaryOp))
```

Reply:
Good point... I don't like that there is no encompassing type for "operations", and I think we should do something about that 😅  What if we add more non-`CustomOp` operations later? That said, these should only be diagonalizing gates, so I think we're already pretty safe in this case. I'll add these to be thorough and call it good enough for now 😆

Reply:
I think we can address this by declaring interfaces that the quantum ops can inherit from. A basic example would be having a `GateOp` base class that all gate type ops inherit. And we can do the same for observables and measurements.